pydm/tests/widgets/test_base.py

```diff
@@ -0,0 +1,540 @@
+        The upper or lower limit can be emitted and subsequently read correctly.
+
+    Parameters
+    ----------
+    qtbot : fixture
+        pytest-qt window for widget test
+    signals : fixture
+        The signals fixture, which provides access signals to be bound to the appropriate slots
+    which_limit : str
+        "UPPER" if the new value is intended for the upper limit, "LOWER" for the lower limit
+    new_limit : float
+        The new limit value
+    """
+    pydm_label = PyDMLabel(init_channel="CA://MA_TEST")
+    qtbot.addWidget(pydm_label)
+
+    if which_limit == "UPPER":
+        signals.upper_ctrl_limit_signal[type(new_limit)].connect(pydm_label.upperCtrlLimitChanged)
+        signals.upper_ctrl_limit_signal[type(new_limit)].emit(new_limit)
+
+        assert pydm_label.get_ctrl_limits()[1] == new_limit
+    elif which_limit == "LOWER":
+        signals.lower_ctrl_limit_signal[type(new_limit)].connect(pydm_label.lowerCtrlLimitChanged)
+        signals.lower_ctrl_limit_signal[type(new_limit)].emit(new_limit)
+
+        assert pydm_label.get_ctrl_limits()[0] == new_limit
+
+
+def test_precision_from_pv(qtbot):
+    """
+    Test setting the flag whether the precision is set to the widget from the PV.
+
+    Expectations:
+    The widget can retain the new precision source flag value.
+
+    Parameters
+    ----------
+    qtbot : fixture
+        pytest-qt window for widget test
+    """
+    pydm_label = PyDMLabel()
+    qtbot.addWidget(pydm_label)
+
+    is_precision_from_pv = pydm_label.precisionFromPV
+
+    # Flip the original flag value. Now the widget should contain the opposite value from before
+    pydm_label.precisionFromPV = not is_precision_from_pv
+
+    assert pydm_label.precisionFromPV is not is_precision_from_pv
+
+
+def test_precision(qtbot):
+    """
+    Test setting the precision is set to the widget.
+
+    Expectations:
+    The widget can retain the new precision value.
+
+    Parameters
+    ----------
+    qtbot : fixture
+       pytest-qt window for widget test
+    """
+    pydm_label = PyDMLabel()
+    qtbot.addWidget(pydm_label)
+
+    pydm_label.precision = 3
+    precision = pydm_label.precision
+
+    pydm_label.precision = precision * 4
+    assert pydm_label.precision == precision * 4
+
+
+def test_channels_for_pydmwidget_base(qtbot):
+    """
+    Test the channels population for the widget whose base class PyDMWidget
+
+    Expectations:
+    1. If the widget does not have channels populated, it will be populated with the default channels, containing all
+        the signals pertaining to its base class.
+    2. Otherwise, the widget will contain just the channels from its latest assignment.
+
+    Parameters
+    ----------
+    qtbot : fixture
+        Window for widget testing
+    """
+    pydm_label = PyDMLabel()
+    qtbot.addWidget(pydm_label)
+
+    assert pydm_label._channel is None
+    assert pydm_label._channels is None
+
+    pydm_channels = pydm_label.channels_for_tools()[0]
```

Comment:
This should not be here. For the channels use the `.channels` property.
The `.channels_for_tools` method is a way for users to define which channels to expose for external tools.